src/hotspot/share/logging/logDecorations.hpp

```diff
@@ -27,21 +27,26 @@
 #include "logging/logDecorators.hpp"
 #include "logging/logTagSet.hpp"
 
-// Temporary object containing the necessary data for a log call's decorations (timestamps, etc).
+class outputStream;
+
+// LogDecorations keeps resolved values for decorators, as well as the
+// printing code to print them. The values are resolved at the log site (in the
+// constructor of LogDecorations); the printing happens when the log message is
+// printed. That may happen delayed, and the object may be stored for some time,
+// in the context of asynchronous logging. Therefore size of this object matters.
 class LogDecorations {
- public:
-  static const int DecorationsBufferSize = 256;
- private:
-  char _decorations_buffer[DecorationsBufferSize];
-  char* _decoration_offset[LogDecorators::Count];
-  LogLevelType _level;
-  const LogTagSet& _tagset;
-  static const char* volatile _host_name;
 
-  const char* host_name();
-  void create_decorations(const LogDecorators& decorators);
+  const jlong _millis;            // for "time", "utctime", "timemillis"
+  const jlong _nanos;             // for "timenanos"
+  const double _elapsed_seconds;  // for "uptime", "uptimemillis", "uptimenanos"
+  const int _pid;                 // for "pid"
```

Comment:
Good point. Might even be more because of padding.

Reply:
... hah, wouldn't you know, its down to 48 bytes :-)

BTW I think if we look closer we can probably squeeze this even more.

E.g. _millis and _nanos are kind of the same thing; for _elapsedSeconds a float may be sufficient instead of a double; and tagset could be made a bitmaks (it really really should be one) and passed as 32bit value object instead of as a reference. But I wanted to keep the patch small and concise for now.